File: connectors/connector-basecamp/src/main/java/com/taskadapter/connector/basecamp/BasecampConnector.java
package com.taskadapter.connector.basecamp;

import com.taskadapter.connector.FieldRow;
import com.taskadapter.connector.NewConnector;
import com.taskadapter.connector.basecamp.transport.ObjectAPI;
import com.taskadapter.connector.basecamp.transport.ObjectAPIFactory;
import com.taskadapter.connector.common.TaskSavingUtils;
import com.taskadapter.connector.definition.ProgressMonitor;
import com.taskadapter.connector.definition.SaveResult;
import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.connector.definition.exceptions.CommunicationException;
import com.taskadapter.connector.definition.exceptions.ConnectorException;
import com.taskadapter.core.PreviouslyCreatedTasksResolver;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import com.taskadapter.model.GUser;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.List;

public class BasecampConnector implements NewConnector {
    /**
     * Keep it the same to enable backward compatibility for previously created config files.
     */
    public static final String ID = "Basecamp 2";

    private final BasecampConfig config;
    private final WebConnectorSetup setup;
    private final ObjectAPIFactory factory;
    private final ObjectAPI api;

    public BasecampConnector(BasecampConfig config, WebConnectorSetup setup, ObjectAPIFactory factory) {
        this.config = config;
        this.setup = setup;
        this.factory = factory;
        try {
            api = factory.createObjectAPI(config, setup);
        } catch (ConnectorException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<GTask> loadData() throws ConnectorException {
        BasecampValidator.validateConfigWithException(config);
        var api = factory.createObjectAPI(config, setup);
        var obj = api.getObject("projects/" + config.getProjectKey()
                + "/todolists/" + config.getTodoKey() + ".json");
        var todosobject = JsonUtils.getOptObject("todos", obj);
        if (todosobject == null) {
            return new ArrayList<>();
        }
        var completed = JsonUtils.getOptArray("completed", todosobject);
        var remaining = JsonUtils.getOptArray("remaining", todosobject);
        var res = new ArrayList<GTask>(JsonUtils.genLen(completed) + JsonUtils.genLen(remaining));
        try {
            if (remaining != null) {
                for (int i = 0; i < remaining.length(); i++) {
                    var task = BasecampToGTask.parseTask(remaining.getJSONObject(i));
                    task.setValue(AllFields.doneRatio, (float) 0);
                    res.add(task);
                }
            }
            if (completed != null && config.getLoadCompletedTodos()) {
                for (int i = 0; i < completed.length(); i++) {
                    var task = BasecampToGTask.parseTask(completed.getJSONObject(i));
                    task.setValue(AllFields.doneRatio, 100f);
                    res.add(task);
                }
            }
        } catch (JSONException e) {
            throw new CommunicationException(e);
        }
        return res;
    }

    @Override
    public GTask loadTaskByKey(TaskId id, Iterable<FieldRow<?>> rows) throws ConnectorException {
        BasecampValidator.validateConfigWithException(config);
        var obj = api.getObject("projects/" + config.getProjectKey() + "/todos/" + id.getKey() + ".json");
        return BasecampToGTask.parseTask(obj);
    }

    @Override
    public SaveResult saveData(PreviouslyCreatedTasksResolver previouslyCreatedTasks,
                               List<GTask> tasks,
                               ProgressMonitor monitor,
                               Iterable<FieldRow<?>> fieldRows) {
        try {
            BasecampValidator.validateConfigWithException(config);
            var users = loadUsers();
            var converter = new GTaskToBasecamp(users);
            var saver = new BasecampSaver(api, config);
            var resultBuilder = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, fieldRows,
                    setup.getHost());

            return resultBuilder.getResult();
        } catch (ConnectorException e) {
            return SaveResult.withError(e);
        }
    }

    private List<GUser> loadUsers() throws ConnectorException {
        if (config.isFindUserByName()) {
            var arr = api.getObjects("people.json");
            var users = new ArrayList<GUser>();
            for (int i = 0; i < arr.length(); i++) {
                try {
                    var user = BasecampToGTask.parseUser(arr.getJSONObject(i));
                    users.add(user);
                } catch (JSONException e) {
                    throw new CommunicationException(e);
                }
            }
            return users;
        } else {
            return List.of();
        }
    }
}

File: connectors/connector-basecamp/src/main/scala/com/taskadapter/connector/basecamp/BasecampConnector.scala
package com.taskadapter.connector.basecamp

import java.util

import com.taskadapter.connector.basecamp.transport.ObjectAPIFactory
import com.taskadapter.connector.common.TaskSavingUtils
import com.taskadapter.connector.definition._
import com.taskadapter.connector.definition.exceptions.{CommunicationException, ConnectorException}
import com.taskadapter.connector.{FieldRow, NewConnector}
import com.taskadapter.core.PreviouslyCreatedTasksResolver
import com.taskadapter.model.{DoneRatio, GTask, GUser}
import org.json.JSONException
import scala.collection.JavaConverters._

object BasecampConnector {
  /**
    * Keep it the same to enable backward compatibility for previously created config files.
    */
  var ID = "Basecamp 2"
}

class BasecampConnector(config: BasecampConfig, setup: WebConnectorSetup, factory: ObjectAPIFactory) extends NewConnector {
  private val api = factory.createObjectAPI(config, setup)

  def loadData(): util.List[GTask] = {
    BasecampValidator.validateConfigWithException(config)
    val api = factory.createObjectAPI(config, setup)
    val obj = api.getObject("projects/" + config.getProjectKey + "/todolists/" + config.getTodoKey + ".json")
    val todosobject = JsonUtils.getOptObject("todos", obj)
    if (todosobject == null) return new util.ArrayList[GTask]
    val completed = JsonUtils.getOptArray("completed", todosobject)
    val remaining = JsonUtils.getOptArray("remaining", todosobject)
    val res = new util.ArrayList[GTask](JsonUtils.genLen(completed) + JsonUtils.genLen(remaining))
    try {
      if (remaining != null) {
        for (i <- 0 until remaining.length) {
          val task = BasecampToGTask.parseTask(remaining.getJSONObject(i))
          task.setValue(DoneRatio, java.lang.Float.valueOf(0))
          res.add(task)
        }
      }
      if (completed != null && config.getLoadCompletedTodos) {
        for (i <- 0 until completed.length) {
          val task = BasecampToGTask.parseTask(completed.getJSONObject(i))
          task.setValue(DoneRatio, java.lang.Float.valueOf(100f))
          res.add(task)
        }
      }
    } catch {
      case e: JSONException =>
        throw new CommunicationException(e)
    }
    res
  }

  @throws[ConnectorException]
  override def loadTaskByKey(id: TaskId, rows: java.lang.Iterable[FieldRow[_]]): GTask = {
    BasecampValidator.validateConfigWithException(config)
    val obj = api.getObject("projects/" + config.getProjectKey + "/todos/" + id.key + ".json")
    BasecampToGTask.parseTask(obj)
  }

  @throws[ConnectorException]
  override def saveData(previouslyCreatedTasks: PreviouslyCreatedTasksResolver, tasks: util.List[GTask],
               monitor: ProgressMonitor,
               fieldRows: java.lang.Iterable[FieldRow[_]]): SaveResult = try {
    BasecampValidator.validateConfigWithException(config)
    val users = loadUsers()
    val converter = new GTaskToBasecamp(users)
    val saver = new BasecampSaver(api, config)
    val resultBuilder = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, fieldRows.asScala,
      setup.host)

    resultBuilder.getResult
  }

  @throws[ConnectorException]
  private def loadUsers(): Seq[GUser] = {
    if (config.isFindUserByName) {
      val arr = api.getObjects("people.json")
      val users = new util.ArrayList[GUser]
      for (i <- 0 until arr.length) {
        try {
          val user = BasecampToGTask.parseUser(arr.getJSONObject(i))
          users.add(user)
        } catch {
          case e: JSONException =>
            throw new CommunicationException(e)
        }
      }
      users.asScala
    } else {
      Seq()
    }
  }
}